decomposer.go

```diff
@@ -0,0 +1,111 @@
+// to respect the negative parameter in the non-finite form.
+//
+// Implementations may choose to signal a negative zero or negative NaN, but implementations
+// that do not support these may also ignore the negative zero or negative NaN without error.
+// If an implementation does not support Infinity it may be converted into a NaN without error.
+// If a value is set that is larger then what is supported by an implementation is attempted to
+// be set, an error must be returned.
+// Implementations must return an error if a NaN or Infinity is attempted to be set while neither
+// are supported.
+type decomposer interface {
+	// Decompose returns the internal decimal state into parts.
+	// If the provided buf has sufficient capacity, buf may be returned as the coefficient with
+	// the value set and length set as appropriate.
+	Decompose(buf []byte) (form byte, negative bool, coefficient []byte, exponent int32)
+
+	// Compose sets the internal decimal value from parts. If the value cannot be
+	// represented then an error should be returned.
+	Compose(form byte, negative bool, coefficient []byte, exponent int32) error
+}
+
+// Decompose returns the internal decimal state into parts.
+// If the provided buf has sufficient capacity, buf may be returned as the coefficient with
+// the value set and length set as appropriate.
+func (d *Dec) Decompose(buf []byte) (form byte, negative bool, coefficient []byte, exponent int32) {
+	high, low := d.coef.hi, d.coef.lo
+	negative = high < 0
+	exponent = -int32(d.scale)
+	if negative {
+		high = -high
+	}
+	size := 16
+	if high == 0 {
+		size = 8
+	}
+	if cap(buf) >= size {
+		coefficient = buf[:size]
+	} else {
+		coefficient = make([]byte, size)
+	}
+	binary.LittleEndian.PutUint64(coefficient, low)
+	if high != 0 {
+		binary.LittleEndian.PutUint64(coefficient[8:], uint64(high))
+	}
+	return form, negative, coefficient, exponent
+}
+
+// Compose sets the internal decimal value from parts. If the value cannot be
+// represented then an error should be returned.
+func (d *Dec) Compose(form byte, negative bool, coefficient []byte, exponent int32) (err error) {
+	if d == nil {
+		return errors.New("Dec must not be nil")
+	}
+	if form != 0 {
+		return errors.New("invalid form, form must be finite")
+	}
+	if exponent > 255 || exponent < -255 {
+		return fmt.Errorf("exponent too large or too small")
+	}
+	var low, high uint64
+	for i, v := range coefficient {
+		if i < 8 {
+			low |= uint64(v) << (i * 8)
+		} else if i < 16 {
+			high |= uint64(v) << ((i - 8) * 8)
+		} else if v != 0 {
+			return fmt.Errorf("coefficent too large")
+		}
+	}
+	if high > math.MaxInt64 {
+		return fmt.Errorf("coefficent too large")
+	}
+	i128 := Int128{
+		hi: int64(high),
+		lo: low,
+	}
+	if negative {
+		i128.hi = -i128.hi
+	}
+	d.coef = i128
+	d.scale = 0
+
+	if exponent > 0 {
+		// Attempt to reduce coefficent. If not possible, return an error.
+		d.coef.Power(intTen, uint(exponent))
```

Comment:
this sets d to 10\*\*exponent
you need to multiply d by 10\*\*exponent
i.e.:
```go
d.coef.Mul(&(d.coef), exp10(exponent))
```

Reply:
Good catch. I've fix this case as well as the Decompose negate issue and added a few more test cases to cover these points.